File: rust/src/seq.rs
pub type Grid = Vec<Vec<u8>>;

pub fn initialize_grid(rows: usize, cols: usize) -> Grid {
    generate_random_grid(rows, cols)
}

pub fn generate_random_grid(rows: usize, cols: usize) -> Grid {
    let mut grid = vec![vec![0; cols]; rows];

    for r in 0..rows {
        for c in 0..cols {
            let value: f32 = rand::random();
            grid[r][c] = if value > 0.7 { 1 } else { 0 };
        }
    }

    grid
}

pub fn next_generation(grid: &Grid) -> Grid {
    let rows = grid.len();
    let cols = grid[0].len();

    let mut new_grid = vec![vec![0; cols]; rows];

    for r in 0..rows {
        for c in 0..cols {
            let neighbors = count_neighbors(grid, r, c);

            new_grid[r][c] = match (grid[r][c], neighbors) {
                (1, x) if x < 2 => 0,
                (1, 2) | (1, 3) => 1,
                (1, x) if x > 3 => 0,
                (0, 3) => 1,
                (state, _) => state,
            };
        }
    }

    new_grid
}

pub fn run_simulation(initial: &Grid, iterations: usize) -> Grid {
    let mut grid = initial.clone();
    for _ in 0..iterations {
        grid = next_generation(&grid);
    }
    grid
}

fn count_neighbors(grid: &Grid, r: usize, c: usize) -> u8 {
    let rows = grid.len() as isize;
    let cols = grid[0].len() as isize;
    let mut count = 0;

    for dr in -1..=1 {
        for dc in -1..=1 {
            if dr == 0 && dc == 0 {
                continue;
            }

            let nr = r as isize + dr;
            let nc = c as isize + dc;

            if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
                count += grid[nr as usize][nc as usize];
            }
        }
    }

    count
}

File: python/src/seq.py
import numpy as np
import os
import time

# NumPy
def initialize_grid(rows, cols, seed=None):
    if seed is not None:
        np.random.seed(seed)    # random 1/0
    # priblizno 30% zivih celija
    # pretvara True/False u 0/1
    return (np.random.rand(rows, cols) > 0.7).astype(np.uint8)

# ne koristim np.roll jer on pravi kopije matrice
def count_neighbors(grid):
    # prazna matrica za broj suseda
    neighbors = np.zeros_like(grid, dtype=np.uint8)

    # 8 suseda
    neighbors[:-1, :] += grid[1:, :]
    neighbors[1:, :] += grid[:-1, :]
    neighbors[:, :-1] += grid[:, 1:]
    neighbors[:, 1:] += grid[:, :-1]

    neighbors[:-1, :-1] += grid[1:, 1:]
    neighbors[:-1, 1:] += grid[1:, :-1]
    neighbors[1:, :-1] += grid[:-1, 1:]
    neighbors[1:, 1:] += grid[:-1, :-1]

    return neighbors

def next_generation(grid):
    neighbors = count_neighbors(grid)

    # pravila (vektorizovana)
    # NumPy koristi C backend, pa se Python interpreter minimalno koristi; mnogo brze od klasicne petlje
    return (
        ((grid == 1) & ((neighbors == 2) | (neighbors == 3))) |
        ((grid == 0) & (neighbors == 3))
    ).astype(np.uint8)

# cuva stanja u datoteku states
def save_grid(grid, iteration, output_dir):
    os.makedirs(output_dir, exist_ok=True)
    filename = os.path.join(output_dir, f"state_{iteration:04}.txt")
    np.savetxt(filename, grid, fmt="%d")

# svaka iteracija koristi prethodno stanje, generise novo i snima rezultat
def run_simulation(rows, cols, iterations, output_dir, seed=None):
    grid = initialize_grid(rows, cols, seed)

    save_grid(grid, 0, output_dir)

    start = time.time()

    for step in range(1, iterations + 1):
        grid = next_generation(grid)
        save_grid(grid, step, output_dir)

    end = time.time()

    print("Sequential Python finished.")
    print(f"Time: {end - start:.6f} s")
